sycl/doc/extensions/proposed/sycl_ext_oneapi_launch_queries.asciidoc

```diff
@@ -0,0 +1,259 @@
+cross-work-group synchronization routines), or when a kernel uses certain
+features (such as static work-group local memory or group algorithms).
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_LAUNCH_QUERIES` to one of the values defined in the
+table below.  Applications can test for the existence of this macro to
+determine if the implementation supports this feature, or applications can test
+the macro's value to determine which of the extension's features the
+implementation supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+
+=== Launch queries
+
+An implementation's ability to satisify requests for specific behaviors
+(such as strong forward progress guarantees and/or cross-work-group
+synchronization) for a kernel may be dependent upon: the kernel itself, and the
+features that it uses; the queue (and underlying device) to which the kernel is
+submitted; and the kernel's launch configuration.
+
+It is a user's responsibility to ensure that a kernel requesting specific
+behaviors uses a compatible launch configuration, using the
+`ext_oneapi_get_info` function and descriptors from the `info::kernel`
+namespace.
+
+If the `sycl::nd_range` parameter used to launch a kernel is incompatible with
+the results of a kernel's launch queries, an implementation must throw a
+synchronous exception with the `errc::nd_range` error code.
+
+[NOTE]
+====
+The values returned by `ext_oneapi_get_info` account for all properties
+attached to a kernel (via the mechanisms defined in the
+sycl_ext_oneapi_kernel_properties extension), as well as the usage of features
+like group algorithms and work-group local memory. Developers should assume
+that the values will differ across kernels.
+====
+
+[source,c++]
+----
+namespace sycl {
+
+class kernel {
+ public:
+  template <typename Param, typename... T>
+  /*return-type*/ ext_oneapi_get_info(T... args) const;
+};
+
+}
+----
+
+[source,c++]
+----
+template <typename Param, typename... T>
+/*return-type*/ ext_oneapi_get_info(T... args) const;
+----
+_Constraints_: Available only when the types `+T...+` described by the parameter
+pack match the types defined in the table below.
+
+_Preconditions_: `Param` must be one of the `info::kernel` descriptors defined
+in this extension.
+
+_Returns_: Information about the kernel that applies when the kernel is
+submitted with the configuration described by the parameter pack `+T...+`.
+The return type is defined in the table below.
+
+This extension adds several new queries to this interface, many of which
+already have equivalents in the `kernel_device_specific` or `device`
+namespaces.
+
+NOTE: These queries are queue- and not device-specific because it is
```

Comment:
Does this mean that `kernel::get_info()` can take a `queue` instead of a `device`?

Reply:
Right. The table has some specific queries accepting a `queue` as their first argument, which is how I'd tried to show this originally.  Do you think we need to add some text here saying that queue-specific queries accept a `queue` as their first argument?

Reply:
I see, sorry I missed that as I was reading it in raw text, I think it'll be clear in the rendered form.